toolchain/check/function.cpp

```diff
@@ -191,60 +191,88 @@ auto CheckFunctionTypeMatches(Context& context,
                      context.functions().Get(prev_function_id), substitutions);
 }
 
+// Checks to see if a structurally valid redeclaration is allowed in context.
+// These all still merge.
+static auto CheckIsAllowedRedecl(Context& context, Parse::NodeId node_id,
+                                 SemIR::Function& new_function,
+                                 bool new_is_definition,
+                                 SemIR::Function& prev_function,
+                                 bool prev_is_import) -> void {
+  CARBON_DIAGNOSTIC(FunctionPreviousDecl, Note, "Previously declared here.");
+  if (prev_is_import) {
+    // TODO: Allow non-extern declarations in the same library.
+    if (!new_function.is_extern && !prev_function.is_extern) {
+      CARBON_DIAGNOSTIC(
+          FunctionNonExternRedecl, Error,
+          "Only one library can declare function {0} without `extern`.",
+          SemIR::NameId);
+      context.emitter()
+          .Build(node_id, FunctionNonExternRedecl, prev_function.name_id)
+          .Note(prev_function.decl_id, FunctionPreviousDecl)
+          .Emit();
+      return;
+    }
+  } else {
+    if (!new_is_definition) {
+      CARBON_DIAGNOSTIC(FunctionRedecl, Error,
+                        "Redundant redeclaration of function {0}.",
+                        SemIR::NameId);
+      context.emitter()
+          .Build(node_id, FunctionRedecl, prev_function.name_id)
+          .Note(prev_function.decl_id, FunctionPreviousDecl)
+          .Emit();
+      return;
+    }
+    if (prev_function.definition_id.is_valid()) {
+      CARBON_DIAGNOSTIC(FunctionRedefinition, Error,
+                        "Redefinition of function {0}.", SemIR::NameId);
+      CARBON_DIAGNOSTIC(FunctionPreviousDefinition, Note,
+                        "Previously defined here.");
+      context.emitter()
+          .Build(node_id, FunctionRedefinition, prev_function.name_id)
+          .Note(prev_function.definition_id, FunctionPreviousDefinition)
+          .Emit();
+      return;
+    }
+    if (new_is_definition && prev_function.is_extern) {
```

Comment:
I think we're missing a diagnostic for:
```
fn F();
extern fn F() {}
```
I wonder if perhaps we could split this up as:

1) Diagnose if `prev_function.is_extern != new_function.is_extern`.
2) Diagnose if `new_is_definition && new_function.is_extern`.

(I'm thinking of approaching it this way because, for `impl`s, we can have more than one non-defining declaration in a file (due to `match_first` blocks and `extend impl` declarations), so we'd be able to reach these checks for `impl`s when neither declaration is a definition.)

Reply:
Relevant test:
https://github.com/carbon-language/carbon-lang/blob/trunk/toolchain/check/testdata/function/definition/extern.carbon#L11-15

That's invalid due to the use of `extern` on a definition, unrelated to the merge. Merging assumes valid declarations as input. I don't think there's a code change here: do you specifically want this case tested, or is the point that the second line is already diagnosed sufficient?

Note, I'd expect similar elsewhere: given `extern` is only valid on forward declarations, and a definition might _not_ merge with a declaration, the `extern` definition needs to be diagnosed on its own.

Reply:
Thanks, looks like we already perform my second check elsewhere. I would find this easier to reason about if we checked for `prev_function.is_extern != new_function.is_extern` here: the current check relies on conditions established elsewhere, which for me makes it harder to see why it's sufficient to correctly handle all cases. But I think this is correct as-is.

Perhaps a better question to ask is, for:
```carbon
extern fn F();
fn F() {}
```
... do we want a diagnostic that says "Cannot define `extern` function `F`.", or do we want a diagnostic that says "Redeclaring `extern` function `F` as non-`extern`."?

To me, the latter is better (though it's marginal) because it's describing a simpler problem -- declaration A is extern, declaration B is not -- which we know is invalid without considering whether declaration B is a definition or not.

But I think this is all a fairly weak argument. Feel free to keep the logic as-is; if you do, for consistency with the previous `if` it'd make sense to drop the tautological check of `new_is_definition`:
```suggestion
    if (prev_function.is_extern) {
```

Reply:
Sure, dropping the tautological check, I'd missed that.

Rephrased the diagnostic per request, I agree that's better.

I'm wary of a check because we don't necessarily update the function to mark is_extern when is_extern would be invalid. But adding a comment.